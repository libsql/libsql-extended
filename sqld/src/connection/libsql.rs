use std::path::{Path, PathBuf};
use std::sync::Arc;

use crossbeam::channel::RecvTimeoutError;
use rusqlite::{ErrorCode, OpenFlags, StatementStatus};
use sqld_libsql_bindings::wal_hook::WalMethodsHook;
use tokio::sync::{oneshot, watch};
use tokio::time::{Duration, Instant};
use tracing::warn;

use crate::auth::{Authenticated, Authorized, Permission};
use crate::error::Error;
use crate::libsql_bindings::wal_hook::WalHook;
use crate::query::Query;
use crate::query_analysis::{State, StmtKind};
use crate::query_result_builder::{QueryBuilderConfig, QueryResultBuilder};
use crate::replication::FrameNo;
use crate::stats::Stats;
use crate::Result;

use super::config::DatabaseConfigStore;
use super::program::{Cond, DescribeCol, DescribeParam, DescribeResponse, DescribeResult};
use super::{MakeConnection, Program, Step, TXN_TIMEOUT};

/// Internal message used to communicate between the database thread and the `LibSqlDb` handle.
type ExecCallback = Box<dyn FnOnce(Result<&mut Connection>) -> anyhow::Result<()> + Send + 'static>;

pub struct LibSqlDbFactory<W: WalHook + 'static> {
    db_path: PathBuf,
    hook: &'static WalMethodsHook<W>,
    ctx_builder: Box<dyn Fn() -> W::Context + Sync + Send + 'static>,
    stats: Arc<Stats>,
    config_store: Arc<DatabaseConfigStore>,
    extensions: Arc<[PathBuf]>,
    max_response_size: u64,
    max_total_response_size: u64,
    auto_checkpoint: u32,
    current_frame_no_receiver: watch::Receiver<Option<FrameNo>>,
    /// In wal mode, closing the last database takes time, and causes other databases creation to
    /// return sqlite busy. To mitigate that, we hold on to one connection
    _db: Option<LibSqlConnection>,
}

impl<W: WalHook + 'static> LibSqlDbFactory<W>
where
    W: WalHook + 'static + Sync + Send,
    W::Context: Send + 'static,
{
    #[allow(clippy::too_many_arguments)]
    pub async fn new<F>(
        db_path: PathBuf,
        hook: &'static WalMethodsHook<W>,
        ctx_builder: F,
        stats: Arc<Stats>,
        config_store: Arc<DatabaseConfigStore>,
        extensions: Arc<[PathBuf]>,
        max_response_size: u64,
        max_total_response_size: u64,
        auto_checkpoint: u32,
        current_frame_no_receiver: watch::Receiver<Option<FrameNo>>,
    ) -> Result<Self>
    where
        F: Fn() -> W::Context + Sync + Send + 'static,
    {
        let mut this = Self {
            db_path,
            hook,
            ctx_builder: Box::new(ctx_builder),
            stats,
            config_store,
            extensions,
            max_response_size,
            max_total_response_size,
            auto_checkpoint,
            current_frame_no_receiver,
            _db: None,
        };

        let db = this.try_create_db().await?;
        this._db = Some(db);

        Ok(this)
    }

    /// Tries to create a database, retrying if the database is busy.
    async fn try_create_db(&self) -> Result<LibSqlConnection> {
        // try 100 times to acquire initial db connection.
        let mut retries = 0;
        loop {
            match self.create_database().await {
                Ok(conn) => return Ok(conn),
                Err(
                    err @ Error::RusqliteError(rusqlite::Error::SqliteFailure(
                        rusqlite::ffi::Error {
                            code: ErrorCode::DatabaseBusy,
                            ..
                        },
                        _,
                    )),
                ) => {
                    if retries < 100 {
                        tracing::warn!("Database file is busy, retrying...");
                        retries += 1;
                        tokio::time::sleep(Duration::from_millis(100)).await
                    } else {
                        Err(err)?;
                    }
                }
                Err(e) => Err(e)?,
            }
        }
    }

    async fn create_database(&self) -> Result<LibSqlConnection> {
        LibSqlConnection::new(
            self.db_path.clone(),
            self.extensions.clone(),
            self.hook,
            (self.ctx_builder)(),
            self.stats.clone(),
            self.config_store.clone(),
            QueryBuilderConfig {
                max_size: Some(self.max_response_size),
                max_total_size: Some(self.max_total_response_size),
                auto_checkpoint: self.auto_checkpoint,
            },
            self.current_frame_no_receiver.clone(),
        )
        .await
    }
}

#[async_trait::async_trait]
impl<W> MakeConnection for LibSqlDbFactory<W>
where
    W: WalHook + 'static + Sync + Send,
    W::Context: Send + 'static,
{
    type Connection = LibSqlConnection;

    async fn create(&self) -> Result<Self::Connection, Error> {
        self.create_database().await
    }
}

#[derive(Clone)]
pub struct LibSqlConnection {
    sender: crossbeam::channel::Sender<ExecCallback>,
}

pub fn open_db<'a, W>(
    path: &Path,
    wal_methods: &'static WalMethodsHook<W>,
    hook_ctx: &'a mut W::Context,
    flags: Option<OpenFlags>,
    auto_checkpoint: u32,
) -> Result<sqld_libsql_bindings::Connection<'a>, rusqlite::Error>
where
    W: WalHook,
{
    let flags = flags.unwrap_or(
        OpenFlags::SQLITE_OPEN_READ_WRITE
            | OpenFlags::SQLITE_OPEN_CREATE
            | OpenFlags::SQLITE_OPEN_URI
            | OpenFlags::SQLITE_OPEN_NO_MUTEX,
    );
    sqld_libsql_bindings::Connection::open(path, flags, wal_methods, hook_ctx, auto_checkpoint)
}

impl LibSqlConnection {
    pub async fn new<W>(
        path: impl AsRef<Path> + Send + 'static,
        extensions: Arc<[PathBuf]>,
        wal_hook: &'static WalMethodsHook<W>,
        hook_ctx: W::Context,
        stats: Arc<Stats>,
        config_store: Arc<DatabaseConfigStore>,
        builder_config: QueryBuilderConfig,
        current_frame_no_receiver: watch::Receiver<Option<FrameNo>>,
    ) -> crate::Result<Self>
    where
        W: WalHook,
        W::Context: Send,
    {
        let (sender, receiver) = crossbeam::channel::unbounded::<ExecCallback>();
        let (init_sender, init_receiver) = oneshot::channel();

        crate::BLOCKING_RT.spawn_blocking(move || {
            let mut ctx = hook_ctx;
            let mut connection = match Connection::new(
                path.as_ref(),
                extensions,
                wal_hook,
                &mut ctx,
                stats,
                config_store,
                builder_config,
                current_frame_no_receiver,
            ) {
                Ok(conn) => {
                    let Ok(_) = init_sender.send(Ok(())) else { return };
                    conn
                }
                Err(e) => {
                    let _ = init_sender.send(Err(e));
                    return;
                }
            };

            loop {
                let exec = match connection.timeout_deadline {
                    Some(deadline) => match receiver.recv_deadline(deadline.into()) {
                        Ok(msg) => msg,
                        Err(RecvTimeoutError::Timeout) => {
                            warn!("transaction timed out");
                            connection.rollback();
                            connection.timed_out = true;
                            connection.timeout_deadline = None;
                            continue;
                        }
                        Err(RecvTimeoutError::Disconnected) => break,
                    },
                    None => match receiver.recv() {
                        Ok(msg) => msg,
                        Err(_) => break,
                    },
                };

                let maybe_conn = if !connection.timed_out {
                    Ok(&mut connection)
                } else {
                    Err(Error::LibSqlTxTimeout)
                };

                if exec(maybe_conn).is_err() {
                    tracing::warn!("Database connection closed unexpectedly");
                    return;
                };
            }
        });

        init_receiver.await??;

        Ok(Self { sender })
    }
}

struct Connection<'a> {
    timeout_deadline: Option<Instant>,
    conn: sqld_libsql_bindings::Connection<'a>,
    timed_out: bool,
    stats: Arc<Stats>,
    config_store: Arc<DatabaseConfigStore>,
    builder_config: QueryBuilderConfig,
    current_frame_no_receiver: watch::Receiver<Option<FrameNo>>,
}

impl<'a> Connection<'a> {
    fn new<W: WalHook>(
        path: &Path,
        extensions: Arc<[PathBuf]>,
        wal_methods: &'static WalMethodsHook<W>,
        hook_ctx: &'a mut W::Context,
        stats: Arc<Stats>,
        config_store: Arc<DatabaseConfigStore>,
        builder_config: QueryBuilderConfig,
        current_frame_no_receiver: watch::Receiver<Option<FrameNo>>,
    ) -> Result<Self> {
        let this = Self {
            conn: open_db(
                path,
                wal_methods,
                hook_ctx,
                None,
                builder_config.auto_checkpoint,
            )?,
            timeout_deadline: None,
            timed_out: false,
            stats,
            config_store,
            builder_config,
            current_frame_no_receiver,
        };

        for ext in extensions.iter() {
            unsafe {
                let _guard = rusqlite::LoadExtensionGuard::new(&this.conn).unwrap();
                if let Err(e) = this.conn.load_extension(ext, None) {
                    tracing::error!("failed to load extension: {}", ext.display());
                    Err(e)?;
                }
                tracing::debug!("Loaded extension {}", ext.display());
            }
        }

        Ok(this)
    }

    fn run<B: QueryResultBuilder>(&mut self, pgm: Program, mut builder: B) -> Result<B> {
        let mut results = Vec::with_capacity(pgm.steps.len());

        builder.init(&self.builder_config)?;
        let is_autocommit_before = self.conn.is_autocommit();

        for step in pgm.steps() {
            let res = self.execute_step(step, &results, &mut builder)?;
            results.push(res);
        }

        // A transaction is still open, set up a timeout
        if is_autocommit_before && !self.conn.is_autocommit() {
            self.timeout_deadline = Some(Instant::now() + TXN_TIMEOUT)
        }

        builder.finish(*self.current_frame_no_receiver.borrow_and_update())?;

        Ok(builder)
    }

    fn execute_step(
        &mut self,
        step: &Step,
        results: &[bool],
        builder: &mut impl QueryResultBuilder,
    ) -> Result<bool> {
        builder.begin_step()?;

        let mut enabled = match step.cond.as_ref() {
            Some(cond) => match eval_cond(cond, results, self.is_autocommit()) {
                Ok(enabled) => enabled,
                Err(e) => {
                    builder.step_error(e).unwrap();
                    false
                }
            },
            None => true,
        };

        let (affected_row_count, last_insert_rowid) = if enabled {
            match self.execute_query(&step.query, builder) {
                // builder error interupt the execution of query. we should exit immediately.
                Err(e @ Error::BuilderError(_)) => return Err(e),
                Err(e) => {
                    builder.step_error(e)?;
                    enabled = false;
                    (0, None)
                }
                Ok(x) => x,
            }
        } else {
            (0, None)
        };

        builder.finish_step(affected_row_count, last_insert_rowid)?;

        Ok(enabled)
    }

    fn execute_query(
        &self,
        query: &Query,
        builder: &mut impl QueryResultBuilder,
    ) -> Result<(u64, Option<i64>)> {
        tracing::trace!("executing query: {}", query.stmt.stmt);

        let config = self.config_store.get();
        let blocked = match query.stmt.kind {
            StmtKind::Read | StmtKind::TxnBegin | StmtKind::Other => config.block_reads,
            StmtKind::Write => config.block_reads || config.block_writes,
            StmtKind::TxnEnd => false,
        };
        if blocked {
            return Err(Error::Blocked(config.block_reason.clone()));
        }

        let mut stmt = self.conn.prepare(&query.stmt.stmt)?;

        let cols = stmt.columns();
        let cols_count = cols.len();
        builder.cols_description(cols.iter())?;
        drop(cols);

        query
            .params
            .bind(&mut stmt)
            .map_err(Error::LibSqlInvalidQueryParams)?;

        let mut qresult = stmt.raw_query();
        builder.begin_rows()?;
        while let Some(row) = qresult.next()? {
            builder.begin_row()?;
            for i in 0..cols_count {
                let val = row.get_ref(i)?;
                builder.add_row_value(val)?;
            }
            builder.finish_row()?;
        }

        builder.finish_rows()?;

        // sqlite3_changes() is only modified for INSERT, UPDATE or DELETE; it is not reset for SELECT,
        // but we want to return 0 in that case.
        let affected_row_count = match query.stmt.is_iud {
            true => self.conn.changes(),
            false => 0,
        };

        // sqlite3_last_insert_rowid() only makes sense for INSERTs into a rowid table. we can't detect
        // a rowid table, but at least we can detect an INSERT
        let last_insert_rowid = match query.stmt.is_insert {
            true => Some(self.conn.last_insert_rowid()),
            false => None,
        };

        drop(qresult);

        self.update_stats(&stmt);

        Ok((affected_row_count, last_insert_rowid))
    }

    fn rollback(&self) {
        let _ = self.conn.execute("ROLLBACK", ());
    }

    fn checkpoint(&self) -> Result<()> {
        self.conn
            .query_row("PRAGMA wal_checkpoint(TRUNCATE)", (), |_| Ok(()))?;
        Ok(())
    }

    fn update_stats(&self, stmt: &rusqlite::Statement) {
        let rows_read = stmt.get_status(StatementStatus::RowsRead);
        let rows_written = stmt.get_status(StatementStatus::RowsWritten);
        let rows_read = if rows_read == 0 && rows_written == 0 {
            1
        } else {
            rows_read
        };
        self.stats.inc_rows_read(rows_read as u64);
        self.stats.inc_rows_written(rows_written as u64);
    }

    fn describe(&self, sql: &str) -> DescribeResult {
        let stmt = self.conn.prepare(sql)?;

        let params = (1..=stmt.parameter_count())
            .map(|param_i| {
                let name = stmt.parameter_name(param_i).map(|n| n.into());
                DescribeParam { name }
            })
            .collect();

        let cols = stmt
            .columns()
            .into_iter()
            .map(|col| {
                let name = col.name().into();
                let decltype = col.decl_type().map(|t| t.into());
                DescribeCol { name, decltype }
            })
            .collect();

        let is_explain = stmt.is_explain() != 0;
        let is_readonly = stmt.readonly();
        Ok(DescribeResponse {
            params,
            cols,
            is_explain,
            is_readonly,
        })
    }

    fn is_autocommit(&self) -> bool {
        self.conn.is_autocommit()
    }
}

fn eval_cond(cond: &Cond, results: &[bool], is_autocommit: bool) -> Result<bool> {
    let get_step_res = |step: usize| -> Result<bool> {
        let res = results.get(step).ok_or(Error::InvalidBatchStep(step))?;
        Ok(*res)
    };

    Ok(match cond {
        Cond::Ok { step } => get_step_res(*step)?,
        Cond::Err { step } => !get_step_res(*step)?,
        Cond::Not { cond } => !eval_cond(cond, results, is_autocommit)?,
        Cond::And { conds } => conds.iter().try_fold(true, |x, cond| {
            eval_cond(cond, results, is_autocommit).map(|y| x & y)
        })?,
        Cond::Or { conds } => conds.iter().try_fold(false, |x, cond| {
            eval_cond(cond, results, is_autocommit).map(|y| x | y)
        })?,
        Cond::IsAutocommit => is_autocommit,
    })
}

fn check_program_auth(auth: Authenticated, pgm: &Program) -> Result<()> {
    for step in pgm.steps() {
        let query = &step.query;
        match (query.stmt.kind, &auth) {
            (_, Authenticated::Anonymous) => {
                return Err(Error::NotAuthorized(
                    "anonymous access not allowed".to_string(),
                ));
            }
            (StmtKind::Read, Authenticated::Authorized(_)) => (),
            (StmtKind::TxnBegin, _) | (StmtKind::TxnEnd, _) => (),
            (
                _,
                Authenticated::Authorized(Authorized {
                    permission: Permission::FullAccess,
                    ..
                }),
            ) => (),
            _ => {
                return Err(Error::NotAuthorized(format!(
                    "Current session is not authorized to run: {}",
                    query.stmt.stmt
                )));
            }
        }
    }
    Ok(())
}

fn check_describe_auth(auth: Authenticated) -> Result<()> {
    match auth {
        Authenticated::Anonymous => {
            Err(Error::NotAuthorized("anonymous access not allowed".into()))
        }
        Authenticated::Authorized(_) => Ok(()),
    }
}

#[async_trait::async_trait]
impl super::Connection for LibSqlConnection {
    async fn execute_program<B: QueryResultBuilder>(
        &self,
        pgm: Program,
        auth: Authenticated,
        builder: B,
        _replication_index: Option<FrameNo>,
    ) -> Result<(B, State)> {
        check_program_auth(auth, &pgm)?;
        let (resp, receiver) = oneshot::channel();
        let cb = Box::new(move |maybe_conn: Result<&mut Connection>| {
            let res = maybe_conn.and_then(|c| {
                let b = c.run(pgm, builder)?;
                let state = if c.conn.is_autocommit() {
                    State::Init
                } else {
                    State::Txn
                };

                Ok((b, state))
            });

            if resp.send(res).is_err() {
                anyhow::bail!("connection closed");
            }

            Ok(())
        });

        let _: Result<_, _> = self.sender.send(cb);

        Ok(receiver.await??)
    }

    async fn describe(
        &self,
        sql: String,
        auth: Authenticated,
        _replication_index: Option<FrameNo>,
    ) -> Result<DescribeResult> {
        check_describe_auth(auth)?;
        let (resp, receiver) = oneshot::channel();
        let cb = Box::new(move |maybe_conn: Result<&mut Connection>| {
            let res = maybe_conn.and_then(|c| c.describe(&sql));

            if resp.send(res).is_err() {
                anyhow::bail!("connection closed");
            }

            Ok(())
        });

        let _: Result<_, _> = self.sender.send(cb);

        Ok(receiver.await?)
    }

    async fn is_autocommit(&self) -> Result<bool> {
        let (resp, receiver) = oneshot::channel();
        let cb = Box::new(move |maybe_conn: Result<&mut Connection>| {
            let res = maybe_conn.map(|c| c.is_autocommit());
            if resp.send(res).is_err() {
                anyhow::bail!("connection closed");
            }
            Ok(())
        });

        let _: Result<_, _> = self.sender.send(cb);
        receiver.await?
    }

    async fn checkpoint(&self) -> Result<()> {
        let (resp, receiver) = oneshot::channel();
        let cb = Box::new(move |maybe_conn: Result<&mut Connection>| {
            let res = maybe_conn.and_then(|c| c.checkpoint());
            if resp.send(res).is_err() {
                anyhow::bail!("connection closed");
            }
            Ok(())
        });

        let _: Result<_, _> = self.sender.send(cb);
        receiver.await?
    }
}

#[cfg(test)]
mod test {
    use itertools::Itertools;

    use crate::query_result_builder::{test::test_driver, IgnoreResult};

    use super::*;

    fn setup_test_conn(ctx: &mut ()) -> Connection {
        let mut conn = Connection {
            timeout_deadline: None,
            conn: sqld_libsql_bindings::Connection::test(ctx),
            timed_out: false,
            stats: Arc::new(Stats::default()),
            config_store: Arc::new(DatabaseConfigStore::new_test()),
            builder_config: QueryBuilderConfig::default(),
            current_frame_no_receiver: watch::channel(None).1,
        };

        let stmts = std::iter::once("create table test (x)")
            .chain(std::iter::repeat("insert into test values ('hello world')").take(100))
            .collect_vec();
        conn.run(Program::seq(&stmts), IgnoreResult).unwrap();

        conn
    }

    #[test]
    fn test_libsql_conn_builder_driver() {
        test_driver(1000, |b| {
            let ctx = &mut ();
            let mut conn = setup_test_conn(ctx);
            conn.run(Program::seq(&["select * from test"]), b)
        })
    }
}
