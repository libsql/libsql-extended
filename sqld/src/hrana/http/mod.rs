use anyhow::{Context, Result};
use parking_lot::Mutex;
use serde::{de::DeserializeOwned, Serialize};
use std::collections::HashMap;
use std::sync::Arc;

use crate::auth::Authenticated;
use crate::database::factory::DbFactory;
use super::ProtocolError;

mod proto;
mod request;
mod stream;

pub struct Server {
    db_factory: Arc<dyn DbFactory>,
    streams: Mutex<HashMap<u64, stream::Handle>>,
    baton_key: [u8; 32],
}

// TODO: release streams after 10 seconds of inactivity

#[derive(Debug)]
pub enum Route {
    GetIndex,
    PostPipeline,
}

impl Server {
    pub fn new(db_factory: Arc<dyn DbFactory>) -> Self {
        Self {
            db_factory,
            streams: Mutex::new(HashMap::new()),
            baton_key: rand::random(),
        }
    }

    pub async fn handle(
        &self,
        auth: Authenticated,
        route: Route,
        req: hyper::Request<hyper::Body>,
    ) -> Result<hyper::Response<hyper::Body>> {
        let res = match route {
            Route::GetIndex => handle_index(),
            Route::PostPipeline => handle_pipeline(self, auth, req).await,
        };
        let resp = match res {
            Ok(resp) => resp,
            Err(err) => {
                let proto_err = err.downcast::<ProtocolError>()?;
                hyper::Response::builder()
                    .status(hyper::StatusCode::BAD_REQUEST)
                    .header(hyper::http::header::CONTENT_TYPE, "text/plain")
                    .body(hyper::Body::from(proto_err.to_string()))
                    .unwrap()
            },
        };
        Ok(resp)
    }
}

fn handle_index() -> Result<hyper::Response<hyper::Body>> {
    Ok(hyper::Response::builder()
        .status(hyper::StatusCode::OK)
        .header(hyper::http::header::CONTENT_TYPE, "text/plain")
        .body(hyper::Body::from("Hello, this is HTTP API v2 (Hrana over HTTP)"))
        .unwrap())
}

async fn handle_pipeline(
    server: &Server,
    auth: Authenticated,
    req: hyper::Request<hyper::Body>,
) -> Result<hyper::Response<hyper::Body>> {
    let req_body: proto::PipelineRequestBody = read_request_json(req).await?;
    let stream_guard = stream::acquire(server, req_body.baton.as_deref()).await?;

    let mut results = Vec::with_capacity(req_body.requests.len());
    for request in req_body.requests.into_iter() {
        let result = request::execute(&stream_guard, auth, request).await
            .context("Could not execute a request in pipeline")?;
        results.push(result);
    }

    let resp_body = proto::PipelineResponseBody {
        baton: stream_guard.release(),
        base_url: None, // TODO: take the base_url from a command-line argument
        results,
    };
    Ok(json_response(&resp_body))
}

async fn read_request_json<T: DeserializeOwned>(req: hyper::Request<hyper::Body>) -> Result<T> {
    let req_body = hyper::body::to_bytes(req.into_body()).await
        .context("Could not read request body")?;
    let req_body = serde_json::from_slice(&req_body)
        .map_err(|err| ProtocolError::Deserialize { source: err })
        .context("Could not deserialize JSON request body")?;
    Ok(req_body)
}

fn json_response<T: Serialize>(resp_body: &T) -> hyper::Response<hyper::Body> {
    let resp_body = serde_json::to_vec(resp_body).unwrap();
    hyper::Response::builder()
        .status(hyper::StatusCode::OK)
        .header(hyper::http::header::CONTENT_TYPE, "application/json")
        .body(hyper::Body::from(resp_body))
        .unwrap()
}
