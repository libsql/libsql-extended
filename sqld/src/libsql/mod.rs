#![allow(improper_ctypes)]

pub mod ffi;
#[cfg(feature = "mwal_backend")]
pub mod mwal;
pub mod wal_hook;

use anyhow::ensure;
use rusqlite::Connection;
use std::os::unix::ffi::OsStrExt;

use crate::libsql::{ffi::libsql_wal_methods_register, wal_hook::WalMethodsHook};

use self::{
    ffi::{libsql_wal_methods, libsql_wal_methods_find},
    wal_hook::WalHook,
};

pub struct WalConnection {
    inner: rusqlite::Connection,
}

impl std::ops::Deref for WalConnection {
    type Target = rusqlite::Connection;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl std::ops::Drop for WalConnection {
    fn drop(&mut self) {
        unsafe {
            rusqlite::ffi::sqlite3_close(self.inner.handle());
        }
        let _ = self.inner;
    }
}

extern "C" {
    fn libsql_open(
        filename: *const u8,
        ppdb: *mut *mut rusqlite::ffi::sqlite3,
        flags: std::ffi::c_int,
        vfs: *const u8,
        wal: *const u8,
    ) -> i32;
}

fn get_orig_wal_methods() -> anyhow::Result<*mut libsql_wal_methods> {
    let orig: *mut libsql_wal_methods = unsafe { libsql_wal_methods_find(0) };
    if orig.is_null() {
        anyhow::bail!("no underlying methods");
    }

    Ok(orig)
}

pub(crate) fn open_with_regular_wal(
    path: impl AsRef<std::path::Path>,
    flags: rusqlite::OpenFlags,
    wal_hook: impl WalHook + 'static,
) -> anyhow::Result<WalConnection> {
    unsafe {
        let mut pdb: *mut rusqlite::ffi::sqlite3 = std::ptr::null_mut();
        let ppdb: *mut *mut rusqlite::ffi::sqlite3 = &mut pdb;
        let orig = get_orig_wal_methods()?;
        let wrapped = WalMethodsHook::wrap(orig, wal_hook);
        let res = libsql_wal_methods_register(wrapped);
        ensure!(res == 0, "failed to register WAL methods");

        let open_err = libsql_open(
            path.as_ref().as_os_str().as_bytes().as_ptr(),
            ppdb,
            flags.bits(),
            std::ptr::null(),
            // FIXME: hack
            "bottomless\0".as_ptr() as *const _,
        );
        assert_eq!(open_err, 0);
        let conn = Connection::from_handle(pdb)?;
        conn.pragma_update(None, "journal_mode", "wal")?;
        tracing::trace!(
            "Opening a connection with regular WAL at {}",
            path.as_ref().display()
        );
        Ok(WalConnection { inner: conn })
    }
}
