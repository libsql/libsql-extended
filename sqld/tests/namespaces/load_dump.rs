use std::convert::Infallible;

use hyper::{service::make_service_fn, Response, Body, StatusCode};
use libsql::{Database, Value};
use serde_json::json;
use tempfile::tempdir;
use tower::service_fn;
use turmoil::Builder;

use crate::common::http::Client;
use crate::common::net::{TurmoilAcceptor, TurmoilConnector};
use crate::namespaces::make_primary;

#[test]
#[ignore = "need to remove block in place in load dump, wait for #693"]
fn load_namespace_from_dump_from_url() {
    const DUMP: &str = r#"
        PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    COMMIT;"#;

    let mut sim = Builder::new().build();
    let tmp = tempdir().unwrap();
    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.host("dump-store", || async {
        let incoming = TurmoilAcceptor::bind(([0, 0, 0, 0], 8080)).await?;
        let server = hyper::server::Server::builder(incoming).serve(make_service_fn(|_conn| async {
            Ok::<_, Infallible>(service_fn(|_req| async {
                Ok::<_, Infallible>(Response::new(Body::from(DUMP)))
            }))
        }));

        server.await.unwrap();

        Ok(())
    });

    sim.client("client", async {
        let client = Client::new();
        client.post("http://primary:9090/v1/namespaces/foo/create", json!({ "dump_url": "http://dump-store:8080/"})).await.unwrap().json::<serde_json::Value>().await.unwrap();

        let foo = Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        let mut rows = foo_conn.query("select count(*) from test", ()).await?;
        assert!(matches!(
                rows.next().unwrap().unwrap().get_value(0)?,
                Value::Integer(1)
        ));

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
#[ignore = "need to remove block in place in load dump, wait for #693"]
fn load_namespace_from_dump_from_file() {
    const DUMP: &str = r#"
        PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    COMMIT;"#;

    let mut sim = Builder::new().build();
    let tmp = tempdir().unwrap();
    let tmp_path = tmp.path().to_path_buf();

    std::fs::write(tmp_path.join("dump.sql"), &DUMP).unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();

        // path is not absolute is an error
        let resp = client.post("http://primary:9090/v1/namespaces/foo/create", json!({ "dump_url": "file:dump.sql"})).await.unwrap();
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST);

        // path doesn't exist is an error
        let resp = client.post("http://primary:9090/v1/namespaces/foo/create", json!({ "dump_url": "file:/dump.sql"})).await.unwrap();
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST);

        let resp = client.post("http://primary:9090/v1/namespaces/foo/create", json!({ "dump_url": format!("file:{}", tmp_path.join("dump.sql").display())})).await.unwrap();
        assert_eq!(resp.status(), StatusCode::OK, "{}", resp.json::<serde_json::Value>().await.unwrap_or_default());

        let foo = Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        let mut rows = foo_conn.query("select count(*) from test", ()).await?;
        assert!(matches!(
                rows.next().unwrap().unwrap().get_value(0)?,
                Value::Integer(1)
        ));

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
#[ignore = "need to remove block in place in load dump, wait for #693"]
fn load_namespace_from_no_commit() {
    const DUMP: &str = r#"
    PRAGMA foreign_keys=OFF;
    BEGIN TRANSACTION;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    "#;

    let mut sim = Builder::new().build();
    let tmp = tempdir().unwrap();
    let tmp_path = tmp.path().to_path_buf();

    std::fs::write(tmp_path.join("dump.sql"), &DUMP).unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();
        let resp = client.post("http://primary:9090/v1/namespaces/foo/create", json!({ "dump_url": format!("file:{}", tmp_path.join("dump.sql").display())})).await.unwrap();
        // the dump is malformed
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST, "{}", resp.json::<serde_json::Value>().await.unwrap_or_default());

        // namespace doesn't exist
        let foo = Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        assert!(foo_conn.query("select count(*) from test", ()).await.is_err());

        Ok(())
    });

    sim.run().unwrap();
}

#[test]
#[ignore = "need to remove block in place in load dump, wait for #693"]
fn load_namespace_from_no_txn() {
    const DUMP: &str = r#"
    PRAGMA foreign_keys=OFF;
    CREATE TABLE test (x);
    INSERT INTO test VALUES(42);
    COMMIT;
    "#;

    let mut sim = Builder::new().build();
    let tmp = tempdir().unwrap();
    let tmp_path = tmp.path().to_path_buf();

    std::fs::write(tmp_path.join("dump.sql"), &DUMP).unwrap();

    make_primary(&mut sim, tmp.path().to_path_buf());

    sim.client("client", async move {
        let client = Client::new();
        let resp = client.post("http://primary:9090/v1/namespaces/foo/create", json!({ "dump_url": format!("file:{}", tmp_path.join("dump.sql").display())})).await.unwrap();
        // the dump is malformed
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST, "{}", resp.json::<serde_json::Value>().await.unwrap_or_default());

        // namespace doesn't exist
        let foo = Database::open_remote_with_connector("http://foo.primary:8080", "", TurmoilConnector)?;
        let foo_conn = foo.connect()?;
        assert!(foo_conn.query("select count(*) from test", ()).await.is_err());

        Ok(())
    });

    sim.run().unwrap();
}
