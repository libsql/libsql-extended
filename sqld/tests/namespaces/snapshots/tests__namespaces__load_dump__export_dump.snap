---
source: sqld/tests/namespaces/load_dump.rs
expression: resp.body_string().await?
---
PRAGMA foreign_keys=OFF;
BEGIN TRANSACTION;
CREATE TABLE IF NOT EXISTS libsql_wasm_func_table (name text PRIMARY KEY, body text) WITHOUT ROWID;
CREATE TABLE IF NOT EXISTS test (x);
INSERT INTO test VALUES(42);
INSERT INTO test VALUES('foo');
INSERT INTO test VALUES('bar');
COMMIT;

